### ISSUE TITLE
[RFC] [Kernel] Add the Kernel component

### ISSUE THREAD

**Issue by alquerci** [symfony/symfony#29275]
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes on version 5.1
| Fixed tickets | #27479
| License       | MIT

Goal
----

Be able to provide a very tiny application with the power of the DIC.

Description
-----

Add ability to use the Dependency Injection factory without HTTP dependencies.

Currently the `HttpKernel` component depends on a lot of packages. However, as soon as we remove all HTTP features dependencies becomes fewer:

* `symfony/config`
* `symfony/dependency-injection`.

One solution is to create the `symfony/kernel` to ease the decoupling.

For history: a similar work has been done for the `security` component. 

Context
-------

Symfony took the path to be tiny and fast. However, CLI applications still required the `HttpKernel` and all related HTTP dependencies that is completely overkill. And this only to have the ability to build and instantiate the service container.

There are also other projects that wants to use or initiate the migration to Symfony. Where migrating the business logic to use the dependency injection component is one of the first action to do.

What will this component contain?
----------------------------

All not HTTP related features of the `HttpKernel`.

```
src/Symfony/Component/Kernel/
├── Bundle
│   ├── BundleInterface.php
│   └── Bundle.php
├── CacheClearer
│   ├── CacheClearerInterface.php
│   ├── ChainCacheClearer.php
│   └── Psr6CacheClearer.php
├── CacheWarmer
│   ├── CacheWarmerAggregate.php
│   ├── CacheWarmerInterface.php
│   ├── CacheWarmer.php
│   └── WarmableInterface.php
├── Config
│   └── FileLocator.php
├── DependencyInjection
│   ├── AddAnnotatedClassesToCachePass.php
│   ├── ConfigurableExtension.php
│   ├── Extension.php
│   ├── LoggerPass.php
│   ├── MergeExtensionConfigurationPass.php
│   ├── ResettableServicePass.php
│   └── ServicesResetter.php
├── EventListener
│   └── DumpListener.php
├── KernelInterface.php
├── Kernel.php
├── Log
│   └── Logger.php
└── RebootableInterface.php
```

Road-map (draft)
-----

- [ ] Extract the kernel from the HTTP one
    * Provides a back compatibility layer for Symfony4.
    * The `symfony/http-kernel` requires and extends the new `symfony/kernel` component.
- [ ] Extract, migrate, refactor the documentation
- [ ] Trigger deprecation on version 5.1
- [ ] Remove BC layer on version 6.0

### Next (out of topic)

* Touch the `FrameworkBundle` to follow the same path.

Migration path (draft)
------

### Required

To migrate to Symfony6: rename the namespace for classes moved to the new component.

```diff
-namespace Symfony\Component\HttpKernel\*;
+namespace Symfony\Component\Kernel\*;
```

### Optional

* For bundles that do not use HTTP functionality, it will be up to the maintainer to remove the HTTP dependency because the migration will not be required.

Resources
-----

* An experimental repository created there https://github.com/alquerci/kernel.
* The [comparison][0] from the version 4.1.6

cc @stof 

[0]: https://github.com/symfony/symfony/compare/v4.1.6...alquerci:add-kernel-component

**Comment by stof** [comment]
your migration path is incomplete. What is the migration path for bundles needing to support multiple versions of Symfony  (and without getting a deprecation warning, so that they don't get bug reports from all their user requiring them to migrate in a way dropping support for the old version) ?

**Comment by stof** [comment]
And this also miss other parts of the migration. Currently, KernelInterface extends httpKernelInterface. How does the HTTP handling looks like once the kernel is not an HttpKernel anymore ?

**Comment by PabloKowalczyk** [comment]
> Add ability to use the Dependency Injection factory outside of HTTP context.

Did you mean "Add ability to use the Dependency Injection outside of HTTP context." (without the **factory**)?

Why `symfony/config` is a dependency?

BTW i like this RFC very much, thanks @alquerci ! 🎉

**Comment by alquerci** [comment]
@stof 

> What is the migration path for bundles needing to support multiple versions of Symfony?

Do you mean for bundle that want to support multiple **major** versions of Symfony. However, even for minor versions this is a good question. So the road to find a solution is to test it. Can you provide a bundle for which this case will happen.

Deprecation are caught and triggered by the debug component even for third party library? For this case that is not the first deprecation of the Symfony life. But this one have a different impact, is it?

This is for sure something to resolve before going forward.

> How does the HTTP handling looks like once the kernel is not an HttpKernel anymore ?

`HttpKernel\Kernel` will extends `Kernel\Kernel` to provide exactly same features as before.
You can see an example [there][0].

---

@PabloKowalczyk 

> Did you mean "Add ability to use the Dependency Injection outside of HTTP context." (without the factory)?

Symfony already provides the ability to use the DIC without any factory. The main scope of this issue is about the DIC factory.

> Why symfony/config is a dependency?

To use the `ConfigCache` and create the config loader as the DIC required a configuration.

[0]: https://github.com/alquerci/symfony/blob/24ef48d326dc76e7023f61e89a98b0935b220cd6/src/Symfony/Component/HttpKernel/Kernel.php

**Comment by stof** [comment]
> Do you mean for bundles that wants to support multiple **major** versions of Symfony. However, even for minor versions this is a good question. So the road to find a solution is to test it. Can you provide a bundle for which this case will happen.

For the ecosystem, we need to allow supporting LTS + next major version at the same time (otherwise bundles will drop support for the LTS, making it useless) and also multiple minor versions at the same time.
And whatever version this changes gets integrated in, some older versions will still be maintained at that time, so it means that bundles will need to be able to support both versions.


> For this case that is not the first deprecation on the Symfony life.

It is indeed not the first one. But we also already rejected some changes in the past because the migration path for the ecosystem was considered too high compared to the benefit. And recently (in #29218), we reverted a deprecation, delaying it to the future, for a similar reason. Considering the cost of the migration, both for projects and for the open-source ecosystem is a necessary part of the decision about deprecating something.

**Comment by alquerci** [comment]
Then if I correctly understood: On version 4.3 adding the component with the BC layer then the deprecation will be triggered on version 5.1 and the BC layer removed for version 6.0.

**Comment by alquerci** [comment]
Hello @Pierstoval 

> With your roadmap, waiting for 6.0 to remove BC layer means that you would have to wait until 2021 for the feature to come... A bit too long, maybe :)

The feature is to extract the DIC factory from the HttpHernel. Then as soon as the new component is released the feature going live too.

What did you expect?

> adding the component for 4.3 and make it not break anything

That's already the plan as you can see on _Resources_ section.

Can you share what will be broken even with the BC layer?

> To me, if the component is accepted by the core team, it needs to be released in 4.3 or 4.4, and old code must be removed in 5.0, in one year

I had thought the same but @stof is right about bundles, see https://github.com/symfony/symfony/issues/29275#issuecomment-440951691 .

---

I am ready to find a solution for each blocker points.

**Comment by alquerci** [comment]
@stof Do you have any news about the status of this feature request ?

Does it have a chance to be added to version 4.3?

cc @fabpot

**Comment by stof** [comment]
Well, the status of this is the same than last time I commented: the suggested migration path cannot be accepted, as it does not account for the migration path for the ecosystem. We need a migration path for the ecosystem (i.e. how the ecosystem can support both a SF version without the new component and a version with the new component).

And btw, this proposal does not clearly describe how the KernelInterface evolves, because that interface currently extends HttpKernelInterface (and the opposite extension cannot be true, as we have a bunch of HttpKernelInterface implementations which are *not* implementing KernelInterface and should not).

**Comment by DanieleGBX** [comment]
Well, we might be able to leverage namespaces for that, `Symfony\Component\HttpKernel\KernelInterface` will become deprecated, and extend both `Symfony\Component\Kernel\KernelInterface` and `Symfony\Component\HttpKernel\HttpKernelInterface`.With `Symfony\Component\Kernel\KernelInterface` describing a "pure" kernel.

**Comment by stof** [comment]
But what will the non-deprecated API look like for a web project be then ?

**Comment by DanieleGBX** [comment]
I was actually discussing this in #general on slack, I'm having a hard time coming up with a future plan, mostly because:

> I am kind of puzzled by what Symfony means by "Kernel". HttpKernelInterface only implements handle, which simply takes a request and returns a response, whereas KernelInterface implements a full system with bundles, resources and whatnot.
> I don't understand how they can both be correct, if they ascribe the same meaning to the term "Kernel".
> I guess, perhaps, if the meaning is exceptionally abstract...
> But at that point, it's kind of a meaningless term, no?

So it all begins with deciding what Kernel means, if it means the core of a Symfony application, then we need a new name for a class that just handles a requests and outputs a response, a name that doesn't include the term Kernel.

**Comment by stof** [comment]
Well, if the kernel is only responsible for managing the bundles and the container, how does the HTTP handling looks like in the front controller ? If the Kernel is not an HttpKernel anymore, it means it cannot work the way it works today. So you need to describe how it would work in your proposal.

**Currently**, a Kernel is managing the bundles and the container, but it is still an HttpKernelInterface implementation.

> then we need a new name for a class that just handles a request and outputs a response, a name that doesn't include the term Kernel.

well, HttpKernelInterface is *only* handling that, nothing else. And even if we rename it to remove `Kernel` from the name, that does not solve the fact that you need to describe how the existing Kernel class gets replaced if it does not implement HttpKernelInterface (or whatever its name).
But anyway, I don't think HttpKernelInterface needs to be renamed.

**Comment by Pierstoval** [comment]
Maybe change the naming, KernelInterface would keep being KernelInterface for apps with bundles etc., and HttpKernelInterface could be renamed with a simple `RequestHandler` or something similar? By the way, this could introduce the concept of middleware or chain of responsibility to Symfony's HTTP handling and make users capable of creating frameworks similar to Express.js with an interface whose name is much clearer about what it does than just HttpKernelInterface, WDYT?

**Comment by stof** [comment]
The naming is not the question here. My question is **What happens for the front controller once the kernel is not able to handle requests anymore ?**. Figuring out the new naming is useless if we don't have the architecture for which we need names yet.

**Comment by DanieleGBX** [comment]
We're not discussing removing the ability for a Kernel to handle requests *in general*, if you have a http application, you're supposed to use a Kernel class that has the handle method, and nothing changes for you.

We're discussing having a Kernel that doesn't have handle methods, for situations where you do *not* have a front controller.

(even tho bin/console would kinda do a quite similar job)

**Comment by Pierstoval** [comment]
Maybe we can make `KernelInterface` not extend `HttpKernelInterface`, and it would be the `src/Kernel.php` that would implement both `KernelInterface` AND `HttpKernelInterface` so it can handle the request, and let the room free for custom application kernels that don't implement `HttpKernelInterface` (such as in 100% command-line applications using kernels)

**Comment by Pierstoval** [comment]
I don't see how HttpKernelInterface would be able to extend KernelInterface since it's _only_ responsible for handling requests, while KernelInterface knows about bundles, config, etc.
On the contrary, the current situation is that the Kernel handles a request and delegates it to a **service**: the HttpKernel, which ONLY handles the request.
Booting application, load cache, build container, etc., these are the roles of a KernelInterface, not an HttpKernelInterface.
That's why I like the idea of removing the "Kernel" part in HttpKernel and split the two concepts to implement them only in the user's kernel. The delegation can still be made to a potential HttpKernelInterface (or RequestHandlerInterface) via the service container, therefore only the abstract `Symfony\Component\HttpKernel\Kernel` would implement the `handle()` method exactly like now. Or maybe we can split this to a trait, like `RequestHandlerKernelTrait` or similar. This splits the logic into different classes. But I'm not sure adding more traits is useful though, I'm not a design expert in this case...

**Comment by DanieleGBX** [comment]
tl:dr;
Now:

- `Symfony\Component\HttpKernel\KernelInterface extends Symfony\Component\HttpKernel\HttpKernelInterface`

Interim solution

- `Symfony\Component\HttpKernel\KernelInterface (deprecated) extends Symfony\Component\Kernel\KernelInterface, Symfony\Component\HttpFoundation\RequestHandlerInterface`
- `Symfony\Component\HttpKernel\HttpKernelInterface (deprecated, alias) extends Symfony\Component\HttpFoundation\RequestHandlerInterface`

Post-deprecation (Symfony 5? 6?)

- `Symfony\Component\HttpKernel\KernelInterface` removed
- `Symfony\Component\HttpKernel\HttpKernelInterface extends Symfony\Component\Kernel\KernelInterface, Symfony\Component\HttpFoundation\RequestHandlerInterface`

**Comment by DanieleGBX** [comment]
Since 4.3 will enter feature freeze soon, should we try to get this added to the list of features before that happens? @alquerci ? @stof ?

**Comment by stof** [comment]
I don't see any reason to rename HttpKernelInterface. That would only make the migration path harder for no reason.

> I agree that the cost for the ecosystem is too high but I do not think higher than for each major Symfony version or bumps the minimal PHP version. Or it will be a good reason to take rid of previous version.

Due to LTS, getting rid of a major version takes years. And that's a **feature** in the ecosystem (otherwise, we would not have a LTS at all)

>     1. Upgrade dependency `symfony/framework-bundle:^5.0`.

That is precisely what we **cannot** enforce as a migration path. That's not a continuous migration path for the ecosystem.

**Comment by alquerci** [comment]
@stof Alright, we need a strong and efficient BC layer.

Write migration tests
----------

We need to write tests for the migration path. With enough tests we will be sure that each steps of the migration works as expected. This also have the advantage to list all cases that are taken account.

The main constraint for the BC layer
----------------

The framework bundle version 3.4 have the following dependency on the http-kernel.

```json
{
    "name": "symfony/framework-bundle",
    "symfony/http-kernel": "~3.4|~4.0",
    "extra": {
        "branch-alias": {
            "dev-master": "3.4-dev"
        }
}
```

So the next dependency will be `~4.4|~5.0`, is it correct?

When a maintainer want to upgrade `http-kernel` to `^5.0`, the upgrade to `framework-bundle:^5.0` will not be required.

### Need help to catch all cases for BC layer

I will not be able to catch all cases alone as this component is part of the foundation of Symfony2+.

Execute tests
---------

As soon as tests are ready they can be executed against the [initial work][0] based on version 4.1.6.

[0]: https://github.com/alquerci/symfony/tree/add-kernel-component

**Comment by DanieleGBX** [comment]
@stof reasons for doing so were provided, quite hard not seeing them *at all.hat said, you might disagree with said reasons, in which case it would be nice to know why, so we can work towards a consensus, as stated it sounds like you're straight up voiding the advocacy efforts for those in favor of the change.

**Comment by stof** [comment]
> When a maintainer want to upgrade `http-kernel` to `^5.0`, the upgrade to `framework-bundle:^5.0` will **not be required**.

We must allow third-party bundles to support `symfony/http-kernel: ~4.4|~5.0`. That's the important part of the migration path. Being able to keep compatibility with both Symfony 4.4 and 5.x at the same time.

**Comment by DanieleGBX** [comment]
@stof, just so I understand, you're saying that a bundle that supports 4.3/4.4 (whenever this feature makes it in) should also support 5.0, provided that the 4.3/4.4 support is without *any* deprecation notices. Correct?

**Comment by stof** [comment]
yes. If you don't have deprecation on 4.4, you are compatible with 5.0. that's the core principle of our deprecation policy.

**Comment by DanieleGBX** [comment]
@alquerci let me know if there's any way I can help to get this out for 4.4.

**Comment by alquerci** [comment]
@DanieleGBX Starting from this week-end I will prepare the PR with the [initial work][0] (rebase and so on).

Todo
----

- [ ] 1. Release specifications see questions just below
- [ ] 2. Create the PR to add the kernel component (@alquerci)
    - [ ] 1. List test cases
    - [ ] 2. Implements tests cases (@alquerci)
    - [ ] 3. Write how to migrate on `UPGRADE_6.0.md`
- [ ] 3. Create the PR for update the documentation

@stof Is something missing?

In short I need help about documentation and list test cases as I may miss some cases.

@DanieleGBX If you are able to write BC layer tests, then I will be happy to have your help too. Also, when the PR will be ready the code review will be next important part.

Questions
-------

Before create the PR I need some feed-backs (specifications).

### A. What is the best implementation for the BC layer?

- [ ] 1. [Class aliases](https://github.com/alquerci/symfony/blob/a80d18ce6d56be843485cbb7236c4b489703df33/src/Symfony/Component/HttpKernel/Resources/bootstrap.php)
- [ ] 2. Standard class extends (where the deprecation will be triggered there on version 6.0)

### B. Do we need to add a BC layer for tests?

- [ ] 1. Yes
- [ ] 2. No

### C. The service resetter implementation have a cross responsibility, so how to split it?

- [ ] 1. [Add scope methods](https://github.com/alquerci/symfony/blob/a80d18ce6d56be843485cbb7236c4b489703df33/src/Symfony/Component/Kernel/Kernel.php#L695-L710)
- [ ] 2. Other

[0]: https://github.com/alquerci/symfony/tree/add-kernel-component

**Comment by alquerci** [comment]
Here we go the PR has been created #30672.

**Comment by fabpot** [comment]
I've just re-read this PR and the related issue. I don't think this is something we will ever do. Lately, we have decoupled things more. And Flex did part of the job. But, having big BC breaks is out of the equation with the maturity of the framework. We might do some small steps in this direction, but no big bang. I'm closing for these reasons.
